middleware.go

```diff
@@ -16,6 +16,27 @@ import (
 
 var defaultMetricPath = "/metrics"
 
+/*
+RequestCounterURLLabelMappingFn is a function which can be supplied to the middleware to control
+the cardinality of the request counter's "url" label, which might be required in some contexts.
+For instance, if for a "/customer/:name" route you don't want to generate a time series for every
+possible customer name, you could use this function:
+
+func(c gin.Context) string {
+	url := c.Request.URL.String()
+	for _, p := range c.Params {
+		if p.Key == "name" {
+			url = strings.Replace(url, p.Value, ":name", 1)
+			break
+		}
+	}
+	return url
+}
+
+which would map "/customer/alice" and "/customer/bob" to "/customer/:name".
+*/
+type RequestCounterURLLabelMappingFn func(c gin.Context) string
```

Comment:
Can we use `func (c *gin.Context)` here instead? I'm not sure I've seen gin.Context referenced without the `*`

Reply:
We could certainly do but the question why is passed around as a pointer? In this case it does not need to mutate any of the value in the context so it should not be a pointer. It could be that the struct is too large (efficiency). Is the case? Maybe. Then it could be to be for consistency which seems to be your point. As a side note, it would be nice to get the input of the gin author.

There is a nice section, on a similar topic, in the [Golang FAQ](
https://golang.org/doc/faq#methods_on_values_or_pointers).